File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
  public List<String> generateParenthesis(int n) {
        List<String> ll = new ArrayList<>();
        Parenthesis(n, 0, 0, "", ll);
        return ll;
    }

    public static void Parenthesis(int n, int open, int close, String ans, List<String> ll) {
		if (open == n && close == n) {
			//System.out.print(ans + " ");
			ll.add(ans);
			return;
		}
		if (open > n || close > open) {
			return;
		}
		Parenthesis(n, open + 1, close, ans + "(", ll);
		Parenthesis(n, open, close + 1, ans + ")", ll);
	}
}

File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let results = [];
    let current = "";
    generate(0, 0, current, n, results);
    return results;
};

function generate(open, close, current, n, results) {
    if(current.length === 2*n) {
        results.push(current);
        return;
    }

    if(open < n) {
        generate(open + 1, close, current + "(", n, results);
    }

    if(close < open) {
        generate(open, close + 1, current + ")", n, results);
    }
}